File: python/CHM.py
"""Microsoft HTML Help."""


import os, re, sys, HTMLParser
import Book


class HHCParser(HTMLParser.HTMLParser):

	def __init__(self, book):
		HTMLParser.HTMLParser.__init__(self)

		self.book = book
		self.contents_stack = []
		self.node = None

	def handle_starttag(self, tag, attrs):
		attrs = dict(attrs)
		if tag == 'ul':
			if len(self.contents_stack) == 0:
				node = self.book.contents
			else:
				assert self.node is not None
				
				node = self.node
				
			self.contents_stack.append(node)
			self.node = None
		elif tag == 'object':
			if attrs['type'] == 'text/sitemap':
				self.node = Book.ContentsNode(None, None)
			else:
				self.node = None
		elif tag == 'param':
			if self.node:
				if attrs['name'] == 'Name':
					self.node.name = attrs['value'].strip()
				elif attrs['name'] == 'Local':
					self.node.link = attrs['value']
			
	def handle_endtag(self, tag):
		if tag == 'ul':
			self.contents_stack.pop()
		elif tag == 'object':
			if self.node:
				self.contents_stack[-1].childs.append(self.node)

	def parse(self, fp):
		self.feed(fp.read())


class HHKParser(HTMLParser.HTMLParser):

	def __init__(self, book):
		HTMLParser.HTMLParser.__init__(self)

		self.book = book
		self.entry = None

	def handle_starttag(self, tag, attrs):
		attrs = dict(attrs)
		if tag == 'object':
			if attrs['type'] == 'text/sitemap':
				self.entry = Book.IndexEntry(None, None)
		elif tag == 'param':
			if self.entry:
				if attrs['name'] == 'Name':
					if self.entry.term is None:
						self.entry.term = attrs['value'].strip()
				elif attrs['name'] == 'Local':
					self.entry.link = attrs['value']
	
	def handle_endtag(self, tag):
		if tag == 'object':
			if self.entry:
				self.book.index.append(self.entry)
				self.entry = None

	def parse(self, fp):
		self.feed(fp.read())


class HHPParser:
	
	OPTCRE = re.compile(
		r'(?P<option>[]\-[\w_.*,(){}]+)'      # a lot of stuff found by IvL
		r'[ \t]*=[ \t]*'                      # followed by separator
		r'(?P<value>.*)$'                     # everything up to eol
		)

	def __init__(self, book):
		self.book = book
		self.section = None
	
	def handle_section(self, name):
		print '[%s]' % (name,)
		self.section = name
	
	def handle_option(self, name, value):
		print name+'='+ value
		if self.section == 'OPTIONS':
			if name == 'Contents file':
				parser = HHCParser(self.book)
				parser.parse(self.book.get(value))
			elif name == 'Index file':
				parser = HHKParser(self.book)
				parser.parse(self.book.get(value))
			elif name == 'Title':
				self.book.title = value
			elif name == 'Default topic':
				self.book.default = value
	
	def handle_line(self, line):
		pass
	
	def parse(self, fp):
		section = None
		while 1:
			line = fp.readline()
			if not line:
				break
			
			# strip comments
			i = line.rfind(';')
			if i >= 0:
				line = line[:i]

			# strip whitespace
			line = line.rstrip()

			# is it empty?
			if line == '':
				continue

			# is it a section header?
			if line[:1] == '[' and line[-1:] == ']':
				section = line[1:-1]
				self.handle_section(section)
			# no section header in the file?
			elif section is None:
				continue
			# an option line?
			elif line.find('=') >= 0:
				optname, optval = line.split('=', 1)
				optname = optname.strip()
				optval = optval.strip()
				# allow empty values
				if optval == '""':
					optval = ''

				self.handle_option(optname, optval)
			else:
				self.handle_line(line)


class MSHHBook(Book.Book):
	"""Microsoft HTML Help."""

	pass


class UncompressedMSHHBook(MSHHBook):

	def __init__(self, hhp):
		MSHHBook.__init__(self)

		self.basedir = os.path.dirname(hhp)

		parser = HHPParser(hhp)
		parser.parse(open(hhp))
	
	def get(self, link):
		path = os.path.join(self.basedir, link)

		return open(path)


class CHMBook(MSHHBook):
	"""Windows Compiled HTML Help."""
	
	# TODO: Fill in the rest.

	pass


import zipfile
try:
	from cStringIO import StringIO
except ImportError:
	from StringIO import StringIO


class HTBBook(MSHHBook):
	"""wxWindows HTML Help Book."""

	def __init__(self, htb):
		MSHHBook.__init__(self)

		self.zip = zipfile.ZipFile(htb, "r")

		for name in self.zip.namelist():
			if name[-4:] == '.hhp':
				parser = HHPParser(self)
				parser.parse(self.get(name))
				return

		raise Exception, "project file not found"
	
	def list(self):
		return self.zip.namelist()
		
	def get(self, link):
		fp = StringIO()
		fp.write(self.zip.read(link))
		fp.seek(0)
		return fp


if __name__ == "__main__":
	import sys
	
	for arg in sys.argv:
		if arg[-4:] == '.htb':
			book = HTBBook(arg)
			print book
		elif arg[-4:] == '.hhp':
			book = UncompressedMSHHBook(arg)
			print book

File: pyhtmlhelp/CHM.cpp
"""Compiled HTML Help book."""


import formatter, htmllib, os, re, sys


hhp = ''
hhc = ''
hhk = ''

name = ''
version = ''
title = ''
link = ''

skip = 0


class RootHHCParser(htmllib.HTMLParser):
	done = 0

	def end_object(self):
		if not self.done:
			self.done = 1

	def unknown_starttag(self, tag, _attributes):
		global name, link, title

		if not self.done:
			attributes = dict(_attributes)
			if tag == 'param':
				if attributes['name'] == 'Name':
					if not title:
						title = attributes['value']
				elif attributes['name'] == 'Local':
					if not name:
						name = attributes['value']
						name = name[:name.find('.')] 
					if not link:
						link = attributes['value']
					
	
class HHCParser(htmllib.HTMLParser):
	level = 0
	open = 0
	sitemap = 0

	def start_ul(self, attributes):
		self.level = self.level + 1
		self.open = 0

	def end_ul(self):
		global output, skip
		
		if self.level > skip:
			if self.open:
				output.write('\t' * self.level + '</sub>\n')
		self.level = self.level - 1
		self.open = 1

	def start_object(self, _attributes):
		global output, skip
		
		if self.level > skip:
			attributes = dict(_attributes)
			if attributes['type'] == 'text/sitemap':
				self.sitemap = 1
				if self.open:
					output.write('\t' * self.level + '</sub>\n')
				output.write('\t' * self.level +  '<sub')
				self.open = 1
	
	def end_object(self):
		global output, skip
		
		if self.sitemap:
			output.write('>\n')
			self.sitemap = 0

	def unknown_starttag(self, tag, _attributes):
		global name, link, title, output, skip

		if self.sitemap:
			attributes = dict(_attributes)
			if tag == 'param':
				if attributes['name'] == 'Name':
					output.write(' name="' + attributes['value'] + '"')
				elif attributes['name'] == 'Local':
					output.write(' link="' + attributes['value'] + '"')


class HHKParser(htmllib.HTMLParser):
	sitemap = 0
	name = 0

	def start_object(self, _attributes):
		global output
		
		attributes = dict(_attributes)
		if attributes['type'] == 'text/sitemap':
			self.sitemap = 1
			output.write('\t<function')
			self.name = 0
	
	def end_object(self):
		global output
		
		if self.sitemap:
			output.write('/>\n')
			self.sitemap = 0

	def unknown_starttag(self, tag, _attributes):
		global output
		
		if self.sitemap:
			attributes = dict(_attributes)
			if tag == 'param':
				if attributes['name'] == 'Name' and not self.name:
					output.write(' name="' + attributes['value'] + '"')
					self.name = 1
				elif attributes['name'] == 'Local':
					output.write(' link="' + attributes['value'] + '"')


def parse_hhp():
	global hhp, hhc, hhk, link, name, output, title

	# Regular expressions for parsing section headers and options.
	SECTCRE = re.compile(
		r'\['                                 # [
		r'(?P<header>[^]]+)'                  # very permissive!
		r'\]'                                 # ]
		)
	OPTCRE = re.compile(
		r'(?P<option>[]\-[\w_.*,(){}]+)'      # a lot of stuff found by IvL
		r'[ \t]*=[ \t]*'                      # followed by separator
		r'(?P<value>.*)$'                     # everything up to eol
		)

	fp = open(hhp)

	cursect = None
	optname = None
	while 1:
		line = fp.readline()
		if not line:
			break
		# comment or blank line?
		if line.strip() == '' or line[0] in '#;':
			continue
		if line.split()[0].lower() == 'rem' \
		   and line[0] in "rR":	  # no leading whitespace
			continue
		# a section header or option header?
		else:
			# is it a section header?
			mo = SECTCRE.match(line)
			if mo:
				sectname = mo.group('header')
				cursect = sectname
				# So sections can't start with a continuation line
				optname = None
			# no section header in the file?
			elif cursect is None:
				continue
			# an option line?
			else:
				mo = OPTCRE.match(line)
				if mo:
					optname, optval = mo.group('option', 'value')
					optname = optname.strip()
					optval = optval.strip()
					# allow empty values
					if optval == '""':
						optval = ''

					if cursect == 'OPTIONS':
						if optname == 'Contents file':
							if not hhc:
								hhc = os.path.join(os.path.dirname(hhp), optval)
						elif optname == 'Index file':
							if not hhk:
								hhk = os.path.join(os.path.dirname(hhp), optval)
						elif optname == 'Title':
							if not title:
								title = optval
						elif optname == 'Default topic':
							if not link:
								link = optval
				else:
					if cursect == 'FILES':
						pass

def parse_hhc_root():
	global hhc, skip
	
	if hhc and skip:
		parser = RootHHCParser(formatter.NullFormatter())
		parser.feed(open(hhc).read())
		parser.close()


def parse_hhc():
	global hhc, output
	
	if hhc:
		output.write('<chapters>\n')
		parser = HHCParser(formatter.NullFormatter())
		parser.feed(open(hhc).read())
		parser.close()
		output.write('</chapters>\n\n')


def parse_hhk():
	global hhk, output
	
	if hhk:
		output.write('<functions>\n')
		parser = HHKParser(formatter.NullFormatter())
		parser.feed(open(hhk).read())
		parser.close()
		output.write('</functions>\n\n')
